Add cluster wide directories path

galexrt opened the pull request:
See https://github.com/rook/rook/issues/1256.

dgonzalezruiz commented on Documentation/cluster-crd.md:
```diff
@@ -70,6 +70,7 @@ Below are the settings available, both at the cluster and individual node level,
   - `databaseSizeMB`:  The size in MB of a bluestore database.
   - `walSizeMB`:  The size in MB of a bluestore write ahead log (WAL).
   - `journalSizeMB`:  The size in MB of a filestore journal.
+- `directories`: Cluster level list of directories to use for storage. These values will be set for all nodes that have no `directories` set.
```
Please indent this properly

galexrt replied:
This is the correct indent. `directories` is not on the level of the `storeConfig`.

dgonzalezruiz commented on pkg/operator/osd/osd.go:
```diff
@@ -107,8 +107,15 @@ func (c *Cluster) Start() error {
 
 			resources := k8sutil.MergeResourceRequirements(c.Storage.Nodes[i].Resources, c.resources)
 
+			var directories []Directory
```
This is more like a noob go question. Wouldn't it be possible to delete the initialization of the directories variable above, and just use the `:=` operator ? (which already does the initialization).

galexrt replied:
In the `if` condition it is a new scope, so if I define/initialize the variable in the `if` the variable isn't available outside the `if` condition, that is why I initialize the variable on line 110.

dgonzalezruiz commented:
Looks good to me!

kokhang commented on pkg/operator/osd/osd.go:
```diff
@@ -107,8 +107,15 @@ func (c *Cluster) Start() error {
 			resources := k8sutil.MergeResourceRequirements(c.Storage.Nodes[i].Resources, c.resources)
 
+			var directories []Directory
+			if len(n.Directories) > 0 {
```
+1

jbw976 commented on pkg/operator/osd/spec.go:
```diff
@@ -24,8 +24,9 @@ import (
-	UseAllNodes bool   `json:"useAllNodes,omitempty"`
+	Nodes       []Node      `json:"nodes,omitempty"`
+	UseAllNodes bool        `json:"useAllNodes,omitempty"`
+	Directories []Directory `json:"directories,omitempty"`
```
instead of duplicating this `Directories` field in the `StorageSpec` type, I would look into refactoring it into the `Selection` type: https://github.com/rook/rook/blob/master/pkg/operator/osd/spec.go#L54

That way, both `StorageSpec` and `Node` get it for free and there's no duplication.

jbw976 commented on pkg/operator/osd/osd.go:
```diff
@@ -107,8 +107,15 @@ func (c *Cluster) Start() error {
 			resources := k8sutil.MergeResourceRequirements(c.Storage.Nodes[i].Resources, c.resources)
 
+			var directories []Directory
+			if len(n.Directories) > 0 {
```
this check looks similar to the work being done in the `resolveNode*` functions: https://github.com/rook/rook/blob/master/pkg/operator/osd/storage.go#L70

if you refactor so that the directories field is in the selection struct, then figuring out if the cluster wide or node specific directories should be used fits naturally in the `resolveNode*` functions.  then you wouldn't even have to make this determination here nor pass a directories arg to `makeReplicaSet`, it would have already figured it all out at resolve time a few lines above.

let me know if that's not clear.

galexrt replied:
@jbw976 If I understand you correctly, the `directories` option I introduce here should be moved here: https://github.com/rook/rook/blob/master/pkg/operator/osd/spec.go#L54-L62 ?

jbw976 replied:
yes, i think that including it in the Selection struct makes sense as the directories list is just another way of selecting storage, and then the directories list doesn't have to be declared in multiple places.

jbw976 commented on Documentation/cluster-crd.md:
```diff
@@ -70,6 +70,7 @@ Below are the settings available, both at the cluster and individual node level,
   - `databaseSizeMB`:  The size in MB of a bluestore database.
   - `walSizeMB`:  The size in MB of a bluestore write ahead log (WAL).
   - `journalSizeMB`:  The size in MB of a filestore journal.
+- `directories`: Cluster level list of directories to use for storage. These values will be set for all nodes that have no `directories` set.
```
if you make the refactor that i've proposed, this would just be part of the storage selection settings list, and there would only need to be one mention of it.

galexrt commented:
@travisn I added a test for the directories "override"/inheritance and removed the empty keys in the added Cluster CRD docs section. @jbw976 I refactored it as you proposed. PTAL

kokhang commented:
is there a breaking change or an api change with this? I noticed you moved the `directories` to be part of the storage selection. If so, the pendingRelease doc should track this.

galexrt commented:
@kokhang It is not a breaking change for users. Only us accessing the `directories`, it has moved from `StorageSpec.Directories` to `StorageSpec.Selection.Directories`.

jbw976 commented on pkg/operator/osd/storage.go:
```diff
@@ -94,6 +94,9 @@ func (s *StorageSpec) resolveNodeConfig(node *Node) {
 	resolveInt(&(node.Config.StoreConfig.WalSizeMB), s.Config.StoreConfig.WalSizeMB, 0)
 	resolveInt(&(node.Config.StoreConfig.JournalSizeMB), s.Config.StoreConfig.JournalSizeMB, 0)
 	resolveString(&(node.Config.Location), s.Config.Location, "")
+	if len(node.Directories) == 0 {
```
to be more accurate, this should be in the `resolveNodeSelection` function instead.

jbw976 commented on pkg/operator/osd/storage_test.go:
```diff
@@ -167,3 +170,74 @@ func TestClearUseAllDevices(t *testing.T) {
 	assert.False(t, storageSpec.AnyUseAllDevices())
 }
+
+func TestClusterDirectoriesInherit(t *testing.T) {
```
very nice thoroughness in this test @galexrt

jbw976 commented on Documentation/cluster-crd.md:
```diff
@@ -60,6 +58,8 @@ Below are the settings available, both at the cluster and individual node level,
   - `^s`: Selects all devices that start with `s`
   - `^[^r]`: Selects all devices that do *not* start with `r`
 - `metadataDevice`: Name of a device to use for the metadata of OSDs on each node.  Performance can be improved by using a low latency device (such as SSD or NVMe) as the metadata device, while other spinning platter (HDD) devices on a node are used to store data.
+- `directories`:  A list of directory paths on this node that will be included in the storage cluster.  Note that using two directories on the same physical device can cause a negative performance impact.
```
maybe we can remove the wording "on this node" now because this is relevant to both the cluster level and the node level.

jbw976 commented on Documentation/cluster-crd.md:
```diff
@@ -177,6 +177,43 @@ spec:
+### Storage Configuration: Cluster wide Directories
+
+This example is based up on the [Storage Configuration: Specific devices](#storage-configuration-specific-devices).
+Individual nodes can override the cluster wiede specified directories list.
```
typo: "wiede".  that looks like a German word :)

galexrt replied:
`wieder` would be German, should have been `wide` ;)

jbw976 commented:
@travisn I would prefer to not change the directories type to just a plain string.  I don't think that would improve the readability of the yaml greatly, it would be a breaking change for users, and it would shut the door for being able to add more properties to dirs later on.

galexrt commented:
@jbw976 @travisn PTAL
